presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java

```diff
@@ -0,0 +1,167 @@
+
+    public ApplyTableScanRedirection(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.domainTranslator = new DomainTranslator(metadata);
+    }
+
+    @Override
+    public Pattern<TableScanNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isTableScanRedirectionEnabled(session);
+    }
+
+    @Override
+    public Result apply(TableScanNode scanNode, Captures captures, Context context)
+    {
+        Optional<TableScanRedirectApplicationResult> tableScanRedirectApplicationResult = metadata.applyTableScanRedirect(context.getSession(), scanNode.getTable());
+        if (tableScanRedirectApplicationResult.isEmpty()) {
+            return Result.empty();
+        }
+
+        CatalogSchemaTableName destinationTable = tableScanRedirectApplicationResult.get().getDestinationTable();
+        Optional<TableHandle> destinationTableHandle = metadata.getTableHandle(
+                context.getSession(),
+                convertFromSchemaTableName(destinationTable.getCatalogName()).apply(destinationTable.getSchemaTableName()));
+        if (destinationTableHandle.isEmpty()) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Destination table %s from table scan redirection not found", destinationTable));
+        }
+        if (destinationTableHandle.get().equals(scanNode.getTable())) {
+            return Result.empty();
+        }
+
+        Map<ColumnHandle, String> columnMapping = tableScanRedirectApplicationResult.get().getDestinationColumns();
+        Map<String, ColumnHandle> destinationColumnHandles = metadata.getColumnHandles(context.getSession(), destinationTableHandle.get());
+        Map<Symbol, ColumnHandle> newScanAssignments = scanNode.getAssignments().entrySet().stream()
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> {
+                    String destinationColumn = requireNonNull(
+                            columnMapping.get(entry.getValue()),
+                            format("Did not find mapping for source column %s in table scan redirection", entry.getValue()));
+                    return requireNonNull(
+                            destinationColumnHandles.get(destinationColumn),
+                            format("Did not find handle for column %s in destination table %s", destinationColumn, destinationTable));
+                }));
+
+        TupleDomain<String> filters = tableScanRedirectApplicationResult.get().getFilters();
+        if (filters.isAll()) {
+            return Result.ofPlanNode(
+                    new TableScanNode(
+                            scanNode.getId(),
+                            destinationTableHandle.get(),
+                            scanNode.getOutputSymbols(),
+                            newScanAssignments,
+                            TupleDomain.all()));
```

Comment:
try to translate `enforcedConstraint` via `TupleDomain#transform` (it works with missing mappings).

You might want to evaluate `Map<ColumnHandle, ColumnHandle> sourceToDestinationColumnHandle` map for this.